src/main/java/com/mmm/his/cer/utility/farser/ast/parser/AstDescentParser.java

```diff
@@ -260,12 +262,61 @@ private <X> Expression<C, X> factor(Expression<C, X> left, int leftOperatorPrece
       this.eat(AstCommonTokenType.RPAREN); // Move iterator if 'RPAREN'
     } else if (commonType == AstCommonTokenType.NOT) {
       left = not(left, leftOperatorPrecedence);
+    } else if (commonType == AstCommonTokenType.FUNCTION) {
+      this.eat();
+      left = createNodeWithArgs();
     } else {
       throw new FarserException("Expression malformed on token " + currentToken);
     }
     return left;
   }
 
+  private CommonTokenFlag getCommonTokenFlag() {
+    return currentToken.getType().getCommonTokenType().orElse(null);
+  }
+
+  /**
+   * Verifies that a token is what is expected. If not, an exception is thrown.
+   *
+   * @param expected The expected token.
+   */
+  private void verifyFunctionTokenValidity(AstCommonTokenType expected) {
+    if (getCommonTokenFlag() != expected) {
+      throw new FarserException("Function expression malformed on token '" + currentToken
+              + "'. Expected '" + expected + "'");
+    }
+  }
+
+  /**
+   * Creates a node with arguments.
+   *
+   * @return The created terminal node.
+   */
+  private <X> Expression<C, X> createNodeWithArgs() {
+    final L functionToken = currentToken; // save the function name
+    this.eat();
+    verifyFunctionTokenValidity(AstCommonTokenType.LPAREN);
+    this.eat();
+
+    CommonTokenFlag commonType = getCommonTokenFlag();
+    List<L> args = new ArrayList<>();
+
+    while (this.tokenIterator.hasNext()
+            && commonType != AstCommonTokenType.RPAREN
+            && commonType != AstCommonTokenType.LPAREN) {
```

Comment:
I added it just so it will fail if you try adding parenthesis within a function call. I am not sure how much we should be error handling malformed formulas in here

Reply:
Makes sense. Maybe add a comment which notes that it is only for validation and not relevant to the actual parsing.